### ISSUE TITLE
Subtyping check failed; Expected type X; got type X

### ISSUE THREAD

**Issue by catalin-hritcu** [FStarLang/FStar#190]
Simon and Enrico got stuck on proving the consequence rule of Hoare logic, because of a problem in F*. I've minimized the example to the following:
```
type form =
| FForall : (heap -> Tot form) -> form

type deduce : form -> Type =
  | DForallElim :
             f:(heap->Tot form) ->
             deduce (FForall f) ->
             x:heap ->
             deduce (f x)

type pred = heap -> Tot form

val hoare_consequence : p:pred -> deduce (FForall (fun h -> p h)) -> h:heap -> Tot unit
let hoare_consequence p vpp' h0 = 
  ignore(DForallElim (fun (h : heap) -> p h) vpp' h0)
```
All we're trying to do is to instantiate a (HOAS represented) quantifier. The error message we get is super strange:
```
Subtyping check failed; expected type (deduce (FForall ((fun h -> (p h)):(heap -> Tot form)))); got type (deduce (FForall ((fun h -> (p h)):(heap -> Tot form))))
```
If I eta contract `(fun g -> p h)` to just `p` then everything works, but in the original example this eta contraction is not possible.

**Comment by catalin-hritcu** [comment]
Generic workaround suggested by Nik is to hoist the function to the top-level, closure converting it and then applying it as needed. This indeed seems to help. Even more generally, point-free reasoning and naming things seems to help work around such problems.

**Comment by nikswamy** [comment]
No longer need to hoist

**Comment by nikswamy** [comment]
re-opening this. the bug is revived by a fix to #260

**Comment by catalin-hritcu** [comment]
There are still some warnings but this works now:
```
../../bin/fstar.exe bug190.fst 
bug190.fst(16,22-16,23) (Warning): Losing precision when encoding a function literal
bug190.fst(16,22-16,23) (Warning): Losing precision when encoding a function literal
Verifying module: Prims
Verifying module: Set
Verifying module: Heap
Verifying module: ST
Verifying module: All
Verifying module: Bug190
All verification conditions discharged successfully
```
